examples/auctions/simple_open_auction.v.py

```diff
@@ -0,0 +1,76 @@
+# Open Auction
+
+# Auction params
+# Beneficiary recieves money from the highest bidder
+beneficiary: public(address)
+auction_start: public(timestamp)
+bidding_time: public(num(sec))
+
+# Current state of auction
+highest_bidder: public(address)
+highest_bid: public(num(wei))
+
+# Allowed withdrawals of previous bids
+pending_returns:(num(wei)[address])
+
+# Set to true at the end, disallows any change
+ended: public(bool)
+
+# Create a simple auction with `_bidding_time`
+# seconds bidding time on behalf of the
+# beneficiary address `_beneficiary`.
+def __init__(_beneficiary: address, _bidding_time: num(sec)):
+    self.beneficiary = _beneficiary
+    self.bidding_time = _bidding_time
+    self.auction_start = block.timestamp
+
+# Bid on the auction with the value sent
+# together with this transaction.
+# The value will only be refunded if the
+# auction is not won.
+@payable
+def bid():
+    # Check if bidding period is over.
+    assert block.timestamp < (self.auction_start)+ self.bidding_time
+    # Check if bid is high enough
+    assert msg.value > self.highest_bid
+    if not self.highest_bid == 0:
+        # Allows the previous highest bidder to take their money back
+        self.pending_returns[self.highest_bidder] += self.highest_bid
+    self.highest_bidder = msg.sender
+    self.highest_bid = msg.value
+
+# Withdraw a bid that was overbid
+def withdraw():
+    amount = self.pending_returns[msg.sender]
+    assert amount > 0
+    self.pending_returns[msg.sender] = 0
+    send(msg.sender, amount)
+
+# End the auction and send the highest bid
+# to the beneficiary.
+def auction_end():
+    # It is a good guideline to structure functions that interact
+    # with other contracts (i.e. they call functions or send Ether)
+    # into three phases:
+    # 1. checking conditions
+    # 2. performing actions (potentially changing conditions)
+    # 3. interacting with other contracts
+    # If these phases are mixed up, the other contract could call
+    # back into the current contract and modify the state or cause
+    # effects (Ether payout) to be performed multiple times.
+    # If functions called internally include interaction with external
+    # contracts, they also have to be considered interaction with
+    # external contracts.
+
+    # 1. Conditions
+    # Check if auction endtime has been reached
+    assert block.timestamp >= self.auction_start + self.bidding_time
+    # Check if this function has already been called
+    assert not self.ended
+
+    # 2. Effects
+    self.ended = True
+
+    # 3. Interaction
+    send(self.beneficiary, self.highest_bid)
```

Comment:
This is in effect sending the bidder back what their bid was, instead of transferring ownership of the bid item to the highest bidder. You would have to have some mechanism to assign ownership to an item that is being bid on, e.g. a contract owner or something? Unclear what this might mean in practice.